test/unittests/oom_test.cpp

```diff
@@ -0,0 +1,92 @@
+// Fizzy: A fast WebAssembly interpreter
+// Copyright 2021 The Fizzy Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "execute.hpp"
+#include "parser.hpp"
+#include <gtest/gtest.h>
+#include <test/utils/asserts.hpp>
+#include <test/utils/execute_helpers.hpp>
+#include <test/utils/hex.hpp>
+
+#ifdef __linux__
+#include <sys/resource.h>
+#endif
+
+#ifndef __has_feature
+#define __has_feature(X) 0
+#endif
+
+#if defined(__SANITIZE_ADDRESS__) || __has_feature(address_sanitizer)
+#define ASAN 1
+#else
+#define ASAN 0
+#endif
+
+namespace
+{
+constexpr uint32_t OSMemoryLimit = 1024 * 1024;
+
+bool try_set_memory_limit(size_t size) noexcept;
+bool restore_memory_limit() noexcept;
+
+#if !ASAN && defined(__linux__) && (defined(__x86_64__) || defined(__i386__))
+
+rlimit orig_limit;
+
+bool try_set_memory_limit(size_t size) noexcept
+{
+    [[maybe_unused]] int err;
+    err = getrlimit(RLIMIT_AS, &orig_limit);
```

Comment:
I wonder how `RLIMIT_AS` is different from `RLIMIT_DATA` (the latter being available on macos too):
```
       RLIMIT_AS
              This is the maximum size of the process's virtual memory
              (address space).  The limit is specified in bytes, and is
              rounded down to the system page size.  This limit affects
              calls to brk(2), mmap(2), and mremap(2), which fail with
              the error ENOMEM upon exceeding this limit.  In addition,
              automatic stack expansion fails (and generates a SIGSEGV
              that kills the process if no alternate stack has been made
              available via sigaltstack(2)).  Since the value is a long,
              on machines with a 32-bit long either this limit is at
              most 2 GiB, or this resource is unlimited.
```

```
       RLIMIT_DATA
              This is the maximum size of the process's data segment
              (initialized data, uninitialized data, and heap).  The
              limit is specified in bytes, and is rounded down to the
              system page size.  This limit affects calls to brk(2),
              sbrk(2), and (since Linux 4.7) mmap(2), which fail with
              the error ENOMEM upon encountering the soft limit of this
              resource.
```

Reply:
I don't know.

Reply:
It is `RLIMIT_DATA` but it does not apply to `mmap` on macos. Please test it and report back.

Reply:
I tested this on macOS. Added `__APPLE__` next to `__linux__`. Everything compiles fine, but it does not work.

1. Tried setting not only `cur`, but `max`. No change.
2. Apparently `RLIMIT_AS` is also available on macOS, just not part of the man page. ([ref](https://github.com/apple/darwin-xnu/blob/main/bsd/sys/resource.h#L474)). Tried, compiles, but no change.
3. It seems that setrlimit/getrlimit is properly implemented in macOS ([ref](https://github.com/apple/darwin-xnu/blob/main/bsd/kern/kern_resource.c#L937)), but `RLIMIT_DATA` or `RLIMIT_AS` is actually not used by anything later on. `RLIMIT_FSIZE/CPU/NPROC/NOFILE/CORE/STACKMEMLOCK` are used in places, so those may be enforced.

See also https://stackoverflow.com/a/6169866 for potential solutions.